File: x/adrsimon/mobile/Dust/Services/ScribeRealtimeClient.swift
import Foundation
import os

private let logger = Logger(subsystem: AppConfig.bundleId, category: "Scribe")

/// Single-use ElevenLabs realtime token, minted by our backend so the device never
/// sees the workspace's ElevenLabs API key.
struct TranscribeToken: Decodable {
    let token: String
    let baseUri: String
}

enum TranscribeTokenService {
    static func fetch(workspaceId: String, tokenProvider: TokenProvider) async throws -> TranscribeToken {
        // Response keys are `token` / `baseUri` (no snake_case), so decode verbatim.
        try await APIClient.authenticatedGet(
            AppConfig.Endpoints.transcribeToken(workspaceId: workspaceId),
            tokenProvider: tokenProvider,
            snakeCase: false
        )
    }
}

/// Minimal client for ElevenLabs Scribe v2 realtime speech-to-text over a WebSocket.
/// Mirrors the wire protocol the `@elevenlabs/client` SDK speaks: audio is streamed as
/// base64 `input_audio_chunk` messages and the server replies with `partial_transcript`
/// (live, in-progress) and `committed_transcript` (finalized segment) messages. Commits
/// are automatic via server-side VAD; we send one manual commit on stop to flush the tail.
final class ScribeRealtimeClient {
    var onPartial: ((String) -> Void)?
    var onCommitted: ((String) -> Void)?
    var onError: ((String) -> Void)?

    private let token: String
    private let baseUri: String
    private let sampleRateHz = 16000

    private var task: URLSessionWebSocketTask?
    private var isClosed = false

    init(token: String, baseUri: String) {
        self.token = token
        self.baseUri = baseUri
    }

    func connect() throws {
        guard var components = URLComponents(string: "\(baseUri)/v1/speech-to-text/realtime") else {
            throw APIError.invalidURL
        }
        components.queryItems = [
            URLQueryItem(name: "model_id", value: "scribe_v2_realtime"),
            URLQueryItem(name: "token", value: token),
            URLQueryItem(name: "commit_strategy", value: "vad"),
            URLQueryItem(name: "audio_format", value: "pcm_16000"),
        ]
        guard let url = components.url else { throw APIError.invalidURL }

        let task = URLSession.shared.webSocketTask(with: url)
        self.task = task
        task.resume()
        receiveNext()
    }

    /// Queues a base64-encoded PCM chunk. `send` buffers until the socket finishes opening,
    /// so chunks captured during the handshake are delivered in order.
    func sendAudio(base64: String) {
        send([
            "message_type": "input_audio_chunk",
            "audio_base_64": base64,
            "commit": false,
            "sample_rate": sampleRateHz,
        ])
    }

    /// Flushes any buffered audio and asks the server to emit a final committed transcript.
    func commit() {
        send([
            "message_type": "input_audio_chunk",
            "audio_base_64": "",
            "commit": true,
            "sample_rate": sampleRateHz,
        ])
    }

    func close() {
        isClosed = true
        task?.cancel(with: .normalClosure, reason: nil)
        task = nil
    }

    private func send(_ object: [String: Any]) {
        guard let task, !isClosed,
              let data = try? JSONSerialization.data(withJSONObject: object),
              let json = String(data: data, encoding: .utf8)
        else { return }
        task.send(.string(json)) { error in
            if let error {
                logger.error("Scribe send failed: \(error)")
            }
        }
    }

    private func receiveNext() {
        task?.receive { [weak self] result in
            guard let self, !self.isClosed else { return }
            switch result {
            case let .success(message):
                if case let .string(text) = message {
                    handle(text)
                }
                receiveNext()
            case let .failure(error):
                onError?("Transcription connection lost: \(error.localizedDescription)")
            }
        }
    }

    private func handle(_ text: String) {
        guard let data = text.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = json["message_type"] as? String
        else { return }

        switch type {
        case "partial_transcript":
            if let value = json["text"] as? String { onPartial?(value) }
        case "committed_transcript", "committed_transcript_with_timestamps":
            if let value = json["text"] as? String { onCommitted?(value) }
        case "session_started":
            break
        default:
            // Every error variant (auth_error, quota_exceeded, …) carries an `error` field.
            if let errorMessage = json["error"] as? String {
                onError?(errorMessage)
            }
        }
    }
}

File: x/aubin/android/app/src/main/kotlin/com/dust/mobile/android/audio/ScribeRealtimeClient.kt
package com.dust.mobile.android.audio

import android.util.Base64
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener

internal sealed interface ScribeTranscriptEvent {
    data class Partial(val text: String) : ScribeTranscriptEvent
    data class Committed(val text: String) : ScribeTranscriptEvent
    data class Error(val message: String) : ScribeTranscriptEvent
    data object Ignored : ScribeTranscriptEvent
}

internal fun parseScribeTranscriptEvent(message: String): ScribeTranscriptEvent {
    val payload = runCatching { Json.parseToJsonElement(message).jsonObject }.getOrNull()
        ?: return ScribeTranscriptEvent.Ignored
    val messageType = payload["message_type"]?.jsonPrimitive?.contentOrNull
    val text = payload["text"]?.jsonPrimitive?.contentOrNull.orEmpty()
    return when (messageType) {
        "partial_transcript" -> ScribeTranscriptEvent.Partial(text)
        "committed_transcript", "committed_transcript_with_timestamps" ->
            ScribeTranscriptEvent.Committed(text)
        "session_started" -> ScribeTranscriptEvent.Ignored
        else -> payload["error"]?.jsonPrimitive?.contentOrNull
            ?.let(ScribeTranscriptEvent::Error)
            ?: ScribeTranscriptEvent.Ignored
    }
}

internal fun scribeAudioMessage(audioBase64: String, commit: Boolean): String =
    buildJsonObject {
        put("message_type", "input_audio_chunk")
        put("audio_base_64", audioBase64)
        put("commit", commit)
        put("sample_rate", ScribeRealtimeClient.SAMPLE_RATE_HZ)
    }.toString()

internal fun scribeRealtimeUrl(baseUri: String, token: String): HttpUrl {
    val handshakeBaseUri = baseUri.trimEnd('/')
        .replaceFirst("wss://", "https://")
        .replaceFirst("ws://", "http://")
    return handshakeBaseUri.toHttpUrl().newBuilder()
        .addPathSegments("v1/speech-to-text/realtime")
        .addQueryParameter("model_id", "scribe_v2_realtime")
        .addQueryParameter("token", token)
        .addQueryParameter("commit_strategy", "vad")
        .addQueryParameter("audio_format", "pcm_16000")
        .build()
}

internal class ScribeRealtimeClient(
    token: String,
    baseUri: String,
    private val httpClient: OkHttpClient = sharedHttpClient,
    private val onEvent: (ScribeTranscriptEvent) -> Unit,
) {
    private val request = Request.Builder()
        .url(scribeRealtimeUrl(baseUri, token))
        .build()

    @Volatile
    private var isClosed = false
    private var webSocket: WebSocket? = null

    fun connect() {
        webSocket = httpClient.newWebSocket(
            request,
            object : WebSocketListener() {
                override fun onMessage(webSocket: WebSocket, text: String) {
                    if (!isClosed) onEvent(parseScribeTranscriptEvent(text))
                }

                override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                    if (!isClosed) {
                        onEvent(ScribeTranscriptEvent.Error(t.message ?: "Transcription connection lost"))
                    }
                }
            },
        )
    }

    fun sendAudio(pcmData: ByteArray) {
        if (isClosed) return
        val base64 = Base64.encodeToString(pcmData, Base64.NO_WRAP)
        webSocket?.send(scribeAudioMessage(base64, commit = false))
    }

    fun commit() {
        if (!isClosed) webSocket?.send(scribeAudioMessage("", commit = true))
    }

    fun close() {
        isClosed = true
        webSocket?.close(1000, null)
        webSocket = null
    }

    companion object {
        const val SAMPLE_RATE_HZ = 16_000
        private val sharedHttpClient = OkHttpClient()
    }
}
